### ISSUE TITLE
YOLOv5l6 @ 1280*1280 with ONNX inference slow

### ISSUE THREAD

**Issue by zhaojun060708** [ultralytics/yolov5#4725]
## ❔Question

I use YOLOv5l6 with 1280 as input size. I convert the model to onnx, and use detect.py to test inference time. 

But I only got 400ms per image with inference time 398ms and nms time 2ms.

When I set batch size to 4, the total inference time is 1.5s which is linear to batch size.

Is this performance normal? Thank you for suggestions.

## Additional context

torch 1.9.0+cu102 CUDA:0 (Tesla P40, 22919.0625MB)
onnx: 1.7.0
onnxruntime: 1.8.1

**Comment by glenn-jocher** [comment]
@zhaojun060708 results will vary by hardware, software, export format and drivers. I can not comment on your specific configuration as we are not in a position to reproduce your particular setup, however YOLOv5l6 inference at 1280 on V100 profile official times per image at batch size 32 are displayed in README here:

<p align="center"><img width="800" src="https://user-images.githubusercontent.com/26833433/114313216-f0a5e100-9af5-11eb-8445-c682b60da2e3.png"></p>
<details>
  <summary>YOLOv5-P5 640 Figure (click to expand)</summary>

<p align="center"><img width="800" src="https://user-images.githubusercontent.com/26833433/114313219-f1d70e00-9af5-11eb-9973-52b1f98d321a.png"></p>
</details>
<details>
  <summary>Figure Notes (click to expand)</summary>

* GPU Speed measures end-to-end time per image averaged over 5000 COCO val2017 images using a V100 GPU with batch size
  32, and includes image preprocessing, PyTorch FP16 inference, postprocessing and NMS.
* EfficientDet data from [google/automl](https://github.com/google/automl) at batch size 8.
* **Reproduce** by
  `python val.py --task study --data coco.yaml --iou 0.7 --weights yolov5s6.pt yolov5m6.pt yolov5l6.pt yolov5x6.pt`

</details>

### Pretrained Checkpoints

[assets]: https://github.com/ultralytics/yolov5/releases

|Model |size<br><sup>(pixels) |mAP<sup>val<br>0.5:0.95 |mAP<sup>test<br>0.5:0.95 |mAP<sup>val<br>0.5 |Speed<br><sup>V100 (ms) | |params<br><sup>(M) |FLOPs<br><sup>640 (B)
|---                    |---  |---      |---      |---      |---     |---|---   |---
|[YOLOv5s][assets]      |640  |36.7     |36.7     |55.4     |**2.0** |   |7.3   |17.0
|[YOLOv5m][assets]      |640  |44.5     |44.5     |63.1     |2.7     |   |21.4  |51.3
|[YOLOv5l][assets]      |640  |48.2     |48.2     |66.9     |3.8     |   |47.0  |115.4
|[YOLOv5x][assets]      |640  |**50.4** |**50.4** |**68.8** |6.1     |   |87.7  |218.8
|                       |     |         |         |         |        |   |      |
|[YOLOv5s6][assets]     |1280 |43.3     |43.3     |61.9     |**4.3** |   |12.7  |17.4
|[YOLOv5m6][assets]     |1280 |50.5     |50.5     |68.7     |8.4     |   |35.9  |52.4
|[YOLOv5l6][assets]     |1280 |53.4     |53.4     |71.1     |12.3    |   |77.2  |117.7
|[YOLOv5x6][assets]     |1280 |**54.4** |**54.4** |**72.0** |22.4    |   |141.8 |222.9
|                       |     |         |         |         |        |   |      |
|[YOLOv5x6][assets] TTA |1280 |**55.0** |**55.0** |**72.0** |70.8    |   |-     |-

<details>
  <summary>Table Notes (click to expand)</summary>

* AP<sup>test</sup> denotes COCO [test-dev2017](http://cocodataset.org/#upload) server results, all other AP results
  denote val2017 accuracy.
* AP values are for single-model single-scale unless otherwise noted. **Reproduce mAP**
  by `python val.py --data coco.yaml --img 640 --conf 0.001 --iou 0.65`
* Speed<sub>GPU</sub> averaged over 5000 COCO val2017 images using a
  GCP [n1-standard-16](https://cloud.google.com/compute/docs/machine-types#n1_standard_machine_types) V100 instance, and
  includes FP16 inference, postprocessing and NMS. **Reproduce speed**
  by `python val.py --data coco.yaml --img 640 --conf 0.25 --iou 0.45 --half`
* All checkpoints are trained to 300 epochs with default settings and hyperparameters (no autoaugmentation).
* Test Time Augmentation ([TTA](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)) includes reflection and scale
  augmentation. **Reproduce TTA** by `python val.py --data coco.yaml --img 1536 --iou 0.7 --augment`

</details>

**Comment by LaserLV52** [comment]
> ## ❔Question
> I use YOLOv5l6 with 1280 as input size. I convert the model to onnx, and use detect.py to test inference time.
> 
> But I only got 400ms per image with inference time 398ms and nms time 2ms.
> 
> When I set batch size to 4, the total inference time is 1.5s which is linear to batch size.
> 
> Is this performance normal? Thank you for suggestions.
> 
> ## Additional context
> torch 1.9.0+cu102 CUDA:0 (Tesla P40, 22919.0625MB)
> onnx: 1.7.0
> onnxruntime: 1.8.1

Maybe your onnx model is work on CPU?  The offical export.py looks like only support the CPU.
My onnx model with yolov5m with 640*640 image also cost 400ms per image with inference time,(my cpu is i5-10400)
If I use the .pt model in detect.py, it only cost 20ms to inference the same image.

**Comment by zhaojun060708** [comment]
> Maybe your onnx model is work on CPU? The offical export.py looks like only support the CPU.
> My onnx model with yolov5m with 640*640 image also cost 400ms per image with inference time,(my cpu is i5-10400)
> If I use the .pt model in detect.py, it only cost 20ms to inference the same image.

Thank you for your reply.

Have you solved this? the pt model inference is 68ms.

Have you tried to export the pt model to onnx supports GPU?

**Comment by zhaojun060708** [comment]
@LaserLV52 Thank you for your reply.

Have you solved this? the pt model inference is 68ms.

Have you tried to export the pt model to onnx supports GPU?

**Comment by LaserLV52** [comment]
> @LaserLV52 Thank you for your reply.
> 
> Have you solved this? the pt model inference is 68ms.
> 
> Have you tried to export the pt model to onnx supports GPU?

Not yet. It looks like no one has solved this problem. Now, I am trying to export the pretrain model to '.torchscript.pt' and use libtorch to deploy it. '.torchscript.pt' seems can be export and work in GPU model, but I haven't used this method before, I am still exploring.
